# Reto 1: Planteamiento y análisis de un problema de optimización

- Paola Andrea Campiño  202020785
- Isabella Contreras Doria 202011871


# Contexto del Problema:

Las enfermeras que trabajan en la clínica del Río en el área de urgencias, se quejaron de sus horarios. Ellas manifiestan que les gustaría trabajar cinco días consecutivos y tener dos días libres consecutivos cada siete días. La clínica recopiló la demanda de los pacientes por día de la semana y sabe cúantas enfermeras debe haber en el personal cada día a la demanda del servicio. La gerencia de Urgencias en la clínica quiere minimizar la nómina de enfermeras al tiempo que reducir las quejas de las enfermeras sobre
sus horarios. Los niveles de dotación de personal y los gastos salariales previstos se muestran en la tabla 1, así como los posibles turnos durante la semana, organizados en la tabla 2.


# Explicación del problema general con palabras



Objetivo general:
- reducir los gastos salariales y beneficios de enfermería.

Restricciones:
- Hay un número minimo de enfermeras por día.
- Las enfermeras solo pueden tener un turno de dos días libres.




In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import minimize

## 1.1 Datos:

**Tabla de turnos**

| Turno | Días libre     |
|--------|-----------|
| A      | (0, 1)    |
| B      | (1, 2)    |
| C      | (2, 3)    |
| D      | (3, 4)    |
| E      | (4, 5)    |
| F      | (5, 6)    |
| G      | (6, 0)    |

<br>

**Cabe resaltar que:**

| Número | Día de la semana     |
|--------|-----------|
| 1      | Lunes    |
| 2      | Martes    |
| 3      | Miercoles    |
| 4      | jueves    |
| 5      | Viernes    |
| 6      | Sabado    |
| 0      | Domingo    |

<br>

|          | Domingo | Lunes | Martes | Miércoles | Jueves | Viernes | Sábado |
|----------|---------|-------|--------|-----------|--------|---------|--------|
| Enfermeras mínimas | 5       | 4     | 3      | 3         | 3      | 4       | 6      |
| Salario  | 320     | 240   | 240    | 240       | 240    | 240     | 320    |


In [ ]:
turnos =    [(0,1), (1,2),(2,3),(3,4),(4,5),(5,6),(6,0)]
dot_gastos =[[5   ,4  ,3    ,3    ,3  ,4    ,6],
             [320 ,240 ,240 ,240 ,240 ,240, 320 ]]

## 1.2 Función objetivo y Variables de desición:

### 1.2.1 Función objetivo:
Se establece la función objetivo meadiante la siguiente expresión matematica:

<br>


$min( \sum_{i \in turnos}
\sum_{j \in diasSemana}
  x_i [persona] * salario_j[dinero/persona]|j\not = turnos_{i0} \land  j\not = turnos_{i1} )$
<br><br>
**En pocas palabras:**
<br>
minimizar la sumatoria de el número de enfermeras por turno $x_i$ multilicado por el salario de los días laborales $salario_j$

**Cabe resaltar que:**
 -  $x_i$ hace referencia a la cantidad de enfermeras que tienen el turno  $i$.
 - $turnos_{i0}$ es el primer día libre del turno.
 - $turnos_{i1}$ es el segundo día libre del turno.


### 1.2.2 Variables de desición:

Para este problema las variables de desición están relacionadas a la cantidad de enfermeras por Turno de descanso. Por lo que se podría pensar las variables de desción de la siguiente manera:
- X= [x_0,x_1,x_2,x_3,x_4,x_5,x_6]<br>
Así mismo es importante mencionar lo siguiente:

- Turno A: $i= 0$,
- Turno B: $i= 1$
- Turno C: $i= 2$
- Turno D: $i= 3$
- Turno E: $i= 4$
- Turno F: $i= 5$
- Turno G: $i= 6$    



In [ ]:
def funcion_objetivo(X):
    z= 0
    for i in range(0,len(turnos)):
      for j in range(0, len(dot_gastos[0])):
        if(j != turnos[i][0] and j != turnos[i][1]):
          z+= X[i]*dot_gastos[1][j]
    return z


### Descripción del código:
- Se hace un recorrido por todos los turnos $i$. Dentro de ese reccorido se recorre los datos relacionados a los salarios y beneficios de enfermería por día de la semana $j$.


- En caso de llegar un día de la semana $j$, en la que el turno $i$ no trabaje **NO** se contabiliza en la sumatoría de Salarios y beneficios de enfería total $z$, en caso contrario se multiplica por la cantidad total de enfermeras $X_i$ que tienen el turno $i$.

## 1.3 Restricciones:

A continuación una formulación general de la restricción de un número mínimo de enfermeras necesarias por día:


##### **Restricción Lunes:**
$∑_{i \in Turnos} x_i [personas] - x_0 -x_1[personas]-enfermerasNecesarias_1[personas] $


##### **Restricción Martes:**
$∑_{i \in Turnos} x_i [personas] - x_1 [personas]-x_2[personas]-enfermerasNecesariasn_2 [personas]$



##### **Restricción Miercoles:**
$∑_{i \in Turnos} x_i [personas]- x_2 [personas]-x_3[personas]-enfermerasNecesarias_3 [personas] $


##### **Restricción Jueves:**
$∑_{i \in Turnos} x_i [personas]- x_3 [personas]-x_4 [personas]-enfermerasNecesarias_4[personas] $


##### **Restricción Viernes:**
$∑_{i \in Turnos} x_i[personas] - x_4 [personas]-x_5[personas]-enfermerasNecesarias_5[personas] $


##### **Restricción Sabado:**
$∑_{i \in Turnos} x_i[personas] - x_5 [personas]-x_6[personas]-enfermerasNecesarias_6[personas] $


##### **Restricción Domingo:**
$∑_{i \in Turnos} x_i [personas] - x_6 [personas]-x_0[personas]-enfermerasNecesarias_0[personas] $


Cabe resaltar que:

- enfermerasNecesarias$_i$: hace referencia al minimo de enfermeras requeridos por cada día de la semana. Tenga en cuenta que $i$: 0  domingo, 1  lunes, 2 martes , 3 miercoles, 4 jueves , 5 viernes y 6 sabados.


In [ ]:
def enfermera_dia_l(X):
      return sum(X)-X[0]-X[1]-dot_gastos[0][1]


def enfermera_dia_ma(X):
      return sum(X)-X[1]-X[2]-dot_gastos[0][2]


def enfermera_dia_mier(X):
      return sum(X)-X[2]-X[3]-dot_gastos[0][3]

def enfermera_dia_juev(X):
      return sum(X)-X[3]-X[4]-dot_gastos[0][4]

def enfermera_dia_vier(X):
      return sum(X)-X[4]-X[5]-dot_gastos[0][5]

def enfermera_dia_sab(X):
      return sum(X)-X[5]-X[6]-dot_gastos[0][6]

def enfermera_dia_dom(X):
      return sum(X)-X[0]-X[1]-dot_gastos[0][0]


constraints = [{'type': 'ineq', 'fun': enfermera_dia_l},
               {'type': 'ineq', 'fun': enfermera_dia_ma},
               {'type': 'ineq', 'fun': enfermera_dia_mier},
               {'type': 'ineq', 'fun': enfermera_dia_juev},
               {'type': 'ineq', 'fun': enfermera_dia_vier},
               {'type': 'ineq', 'fun': enfermera_dia_sab},
               {'type': 'ineq', 'fun': enfermera_dia_dom}]

### Descripción del código:


- Para cada uno de los días lo que dice es: **La total de enfermeras** - **aquellas que tienen día libre ese día**- **el minimo de enfermeras para ese día** $\geq 0$.
- Esto se establece esto para cada uno de los días de la semana.

## **1.4.** Definir los límites para cada una de las variables (bounds)


- Se sabe que las variable pueden tomar cualquier valor entre los números reales, pero este valor solo puede ser positivo (no hay manera de tener enfermeras negativas).

In [ ]:
initial_guess = [0,0,0,0,0,0,0]
boundaries = [(0, None),(0, None),(0, None),(0, None),(0, None),(0, None),(0, None)]

### Descripción del código:


- se ha establecido los limites para cada variable entre $0\leq n\leq∞+$.
- El valor inicial se ha tomado [0,0,0,0,0,0].

## **1.5** Calculo del minimizador

In [ ]:

resultado = minimize(funcion_objetivo, initial_guess, constraints=constraints,bounds=boundaries)
print(resultado)

 message: Positive directional derivative for linesearch
 success: False
  status: 8
     fun: 8079.959296382605
       x: [ 1.000e+00  0.000e+00  2.000e+00  1.000e+00  2.000e+00
            0.000e+00  0.000e+00]
     nit: 7
     jac: [ 1.280e+03  1.360e+03  1.360e+03  1.360e+03  1.360e+03
            1.280e+03  1.200e+03]
    nfev: 24
    njev: 3


**5.** Resultados

In [ ]:
rounded_solution = [round(val, 2) for val in resultado.x]

print("Valor óptimo de las variables:", rounded_solution)
print("Valor óptimo de la función objetivo:", resultado.fun)

Valor óptimo de las variables: [1.0, 0.0, 2.0, 1.0, 2.0, 0.0, 0.0]
Valor óptimo de la función objetivo: 8079.959296382605


### Descripción del código:

- después de hacer uso de la función minimize se ha llegado a los resultados los cualesson redondeados de manera que se vean de manera clara.
- Así mismo en consola se ve el valor que toma la función objetvio.

## 1.7 Verificación de la solución.

In [ ]:
def check_result(X= rounded_solution):
    todo_correcto = True
    correct_rest = {}
    correct_rest["Lunes"]=enfermera_dia_l(X)
    correct_rest["Martes"]= enfermera_dia_ma(X)
    correct_rest["Miercoles"]=enfermera_dia_mier(X)
    correct_rest["Jueves"]=enfermera_dia_juev(X)
    correct_rest["Viernes"]=enfermera_dia_vier(X)
    correct_rest["Sabados"]=enfermera_dia_sab(X)
    correct_rest["Domingo"]=enfermera_dia_dom(X)
    print(correct_rest)

    for i in correct_rest:
      if correct_rest[i]<0:
        print("No se cumple la restricción para el "+i+" hay solo "+str(correct_rest[i])+" enfermeras, siendo el minimo de"+str(abs(correct_rest[i])))
        return todo_correcto



    todo_correcto= True
    return todo_correcto



print ("El resultado cumple con las restricciones: "+str(check_result()))

{'Lunes': 1.0, 'Martes': 1.0, 'Miercoles': 0.0, 'Jueves': 0.0, 'Viernes': 0.0, 'Sabados': 0.0, 'Domingo': 0.0}
El resultado cumple con las restricciones: True


### Descripción del código:

Para hacer la verificación de los datos resultantes, se ha tomado la respuesta y usado las funciones que modelan la restricciones. Esto se guarda en un diccionario donde el día es la llave y el valor es el resultado de meterlo a la restricción. Si el valor es igual a un numero menor a 0, se puede decir que el resutado está mal pues hay pocas enfermeras disponibles para ese día.

Así mismo la función da una respuesta siendo `True`, cuando el resultado cumple con las restricciones y `False` cuando no se cumple.

## 1.6 Solución al problema



In [ ]:
def print_recurso(solution=rounded_solution):
    print("Nomina total Semanal para todas las enfermeras:")
    print(round(resultado.fun,2))




    print("\nCantidad de enfermeras por turno de días libres:")

    print("Turno A: "+str(solution[0]))
    print("Turno B: "+str(solution[1]))
    print("Turno C: "+str(solution[2]))
    print("Turno D: "+str(solution[3]))
    print("Turno E: "+str(solution[4]))
    print("Turno F: "+str(solution[5]))
    print("Turno G: "+str(solution[6]))


    print("\n")
print_recurso()

Nomina total Semanal para todas las enfermeras:
8079.96

Cantidad de enfermeras por turno de días libres:
Turno A: 1.0
Turno B: 0.0
Turno C: 2.0
Turno D: 1.0
Turno E: 2.0
Turno F: 0.0
Turno G: 0.0




## 2. Análisis de la solución

### 2.1 Prueba de puntos en la región factible:

Tenga en cuenta que la variable objetivo llego a tomar un valor de 8,079.96, con el vector [1,0,2,1,2,0,0].


Se realizó pruebas con diferentes valores que entran dentro de la región factible, y se ha encontrado que ninguno llega a ser menor. A continuación los puntos evaludos:
- [2,1,1,1,1,1,2]


In [ ]:
prueba_1=[2,1,1,1,1,1,2]
print(funcion_objetivo(prueba_1))
check_result(prueba_1)


{'Lunes': 2, 'Martes': 4, 'Miercoles': 4, 'Jueves': 4, 'Viernes': 3, 'Sabados': 0, 'Domingo': 1}


True

In [ ]:
prueba_2= [1,2,1,1,1,1,1]

print(funcion_objetivo(prueba_2))
check_result(prueba_2)

10560
{'Lunes': 1, 'Martes': 2, 'Miercoles': 3, 'Jueves': 3, 'Viernes': 2, 'Sabados': 0, 'Domingo': 0}


True

Cabe resaltar que se ha verificado que está en la región factible por medio de la función check_result en la que se evalua todos los valores para cada una de las restricciones, con lo que podemos notar que al ningun elemento dentro del diccionario estar en número negativo se puede decir que están en la región factible.

### 2.2 Prueba de diferentes puntos inciales:


En los siguientes ejemplos, se logra ver una prueba, la cual confirma que a pesar de tener distintos puntos iniciales, los resultados siguen siendo iguales para esta optimización. En el primer ejemplo se inicia con el punto [1,3,5,4,7,9,2] y en el segundo se inicia con el punto [2,4,6,8,10,12,14]

In [ ]:
initial_guess = [1,3,5,4,7,9,2]
boundaries = [(0, None),(0, None),(0, None),(0, None),(0, None),(0, None),(0, None)]

resultado_prueba = minimize(funcion_objetivo, initial_guess, constraints=constraints,bounds=boundaries)
print(resultado_prueba)

rounded_solution = [round(val, 2) for val in resultado_prueba.x]

print("Valor óptimo de las variables:", rounded_solution)
print("Valor óptimo de la función objetivo:", resultado.fun)

 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 8079.997026400088
       x: [ 1.000e+00  0.000e+00  2.000e+00  1.000e+00  2.000e+00
            0.000e+00  0.000e+00]
     nit: 6
     jac: [ 1.280e+03  1.360e+03  1.360e+03  1.360e+03  1.360e+03
            1.280e+03  1.200e+03]
    nfev: 16
    njev: 2
Valor óptimo de las variables: [1.0, 0.0, 2.0, 1.0, 2.0, 0.0, 0.0]
Valor óptimo de la función objetivo: 8079.997026400088


In [ ]:
initial_guess = [2,4,6,8,10,12,14]
boundaries = [(0, None),(0, None),(0, None),(0, None),(0, None),(0, None),(0, None)]

resultado_prueba = minimize(funcion_objetivo, initial_guess, constraints=constraints,bounds=boundaries)
print(resultado_prueba)

rounded_solution = [round(val, 2) for val in resultado_prueba.x]

print("Valor óptimo de las variables:", rounded_solution)
print("Valor óptimo de la función objetivo:", resultado.fun)

 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 8079.997374994038
       x: [ 1.000e+00  3.901e-08  2.000e+00  1.000e+00  2.000e+00
            3.637e-08  3.261e-08]
     nit: 7
     jac: [ 1.280e+03  1.360e+03  1.360e+03  1.360e+03  1.360e+03
            1.280e+03  1.200e+03]
    nfev: 28
    njev: 3
Valor óptimo de las variables: [1.0, 0.0, 2.0, 1.0, 2.0, 0.0, 0.0]
Valor óptimo de la función objetivo: 8079.997026400088
